### PR TITLE
[octavia-lb] Add the logic to reattach the VIP to the load balancer if the IP has changed

### LINKED ISSUE #2443: [occm] loadbalancer doesn't honor the `spec.loadBalancerIP` on service update
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

Documentation [says](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/openstack-cloud-controller-manager/expose-applications-using-loadbalancer-type-service.md#creating-service-by-specifying-a-floating-ip):

> Sometimes it's useful to use an existing available floating IP rather than creating a new one, especially in the automation scenario. In the example below, 122.112.219.229 is an available floating IP created in the OpenStack Networking service.

>> NOTE: If 122.112.219.229 is not available, a new floating IP will be created automatically from the configured public network. If 122.112.219.229 is already associated with another port, the Service creation will fail.

However once I create a svc with no `spec.loadBalancerIP` defined, OCCM assigns the random FIP to a LB. If I want to change the FIP on the existing service by setting the `spec.loadBalancerIP`, the setting persists, but the actual FIP is not assigned.

**What you expected to happen**:

I expect OCCM to assign a new FIP once `spec.loadBalancerIP` is changed.

**How to reproduce it**:

* create a service without a FIP
* assign a  `spec.loadBalancerIP` to a desired value
* FIP is not changed

**Anything else we need to know?**:

https://github.com/kubernetes/cloud-provider-openstack/blob/ed517e1f07057cd439542c73bff12102d8849dde/pkg/openstack/loadbalancer.go#L1039-L1041

see also #2377, where this bug is seen more clearly.

**Environment**:
- openstack-cloud-controller-manager(or other related binary) version: master
- OpenStack version: ???
- Others: ???

### PR DESCRIPTION
**Pull request by yang-wang11**
**What this PR does / why we need it**:

**Which issue this PR fixes(if applicable)**:
fixes #2443

**Special notes for reviewers**:

**Release note**:

```release-note
Enable OCCM to reassign the new Floating IP from service.spec when spec.loadBalancerIP is changed.
```

### CODE REVIEW

**Review by dulek — CHANGES_REQUESTED** [review]
I think this enables users of the cluster to set `spec.loadBalancerIP=""`, wait for a new FIP to get attached and repeat the process until all FIPs are exhausted. We should allow that. At minimum old FIP has to be deleted if it was created by the CPO.

**Comment by dulek** [comment]
Seems like this also breaks shared LBs:

```
E1024 12:12:45.460335      11 controller.go:298] error processing service octavia-lb-test/test-shared-2 (retrying with exponential backoff): failed to ensure load balancer: floating IP  is not available
I1024 12:12:45.460700      11 event.go:376] "Event occurred" object="octavia-lb-test/test-shared-2" fieldPath="" kind="Service" apiVersion="v1" type="Warning" reason="SyncLoadBalancerFailed" message="Error syncing load balancer: failed to ensure load balancer: floating IP  is not available"
```

Not that I like the feature, but we cannot just break it.

**Comment by yang-wang11** [comment]
> I think this enables users of the cluster to set `spec.loadBalancerIP=""`, wait for a new FIP to get attached and repeat the process until all FIPs are exhausted. We shouldn't allow that. At minimum old FIP has to be deleted if it was created by the CPO.

@dulek  Great observation, I will take some time to refactor this function to lessen the if-else.

File: pkg/openstack/loadbalancer.go [hunk 1]
```diff
@@ -430,6 +430,14 @@ func getSecurityGroupName(service *corev1.Service) string {
 	return securityGroupName
 }
 
+func getLoadBalancerIP(service *corev1.Service) string {
+	return service.Spec.LoadBalancerIP
```
Comment (yang-wang11) [inline pkg/openstack/loadbalancer.go:434]: Should further annotation support be provided, enhancement can be realized here.

Comment (yang-wang11) [reply pkg/openstack/loadbalancer.go]: enhanced in the commit https://github.com/kubernetes/cloud-provider-openstack/pull/2451/commits/58812808fa44205510af62d654c6e4e55c25bf9d.

File: docs/openstack-cloud-controller-manager/expose-applications-using-loadbalancer-type-service.md [hunk 2]
```diff
@@ -303,13 +303,17 @@ spec:
 
 Sometimes it's useful to use an existing available floating IP rather than creating a new one, especially in the automation scenario. In the example below, 122.112.219.229 is an available floating IP created in the OpenStack Networking service.
 
-> NOTE: If 122.112.219.229 is not available, a new floating IP will be created automatically from the configured public network. If 122.112.219.229 is already associated with another port, the Service creation will fail.
+The service’s annotation loadbalancer.openstack.org/floating-ip can also be utilized to specify a desired floating IP. If this annotation exists and is not empty, the value contained in it will be used as the floating IP. Otherwise, the spec.loadBalancerIP field will be used as shown in the example below.
+
```
Comment (jeffyjf) [inline docs/openstack-cloud-controller-manager/expose-applications-using-loadbalancer-type-service.md:307]: The new annotation should be listed [here](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/openstack-cloud-controller-manager/expose-applications-using-loadbalancer-type-service.md?plain=1#L91)

**Review by jeffyjf — CHANGES_REQUESTED** [review]

File: pkg/openstack/loadbalancer.go [hunk 3]
```diff
@@ -944,103 +962,131 @@ func (lbaas *LbaasV2) ensureFloatingIP(clusterName string, service *corev1.Servi
...
+}
+
+func (lbaas *LbaasV2) processWithFloatIP(portID string, floatIP *floatingips.FloatingIP, service *corev1.Service) (string, error) {
+
+	klog.V(4).Infof("Found floating ip %v by loadbalancer port id %q", floatIP, portID)
+
+	loadBalancerIP := getLoadBalancerIP(service)
+
+	if floatIP.FloatingIP != loadBalancerIP && loadBalancerIP != "" {
+		if err := lbaas.associateFloatingIP(loadBalancerIP, portID); err != nil {
```
[the remark this answers is not in the document]
Comment (yang-wang11) [reply pkg/openstack/loadbalancer.go:1062]: yes, the limitation is needed, only the LB-owned can be updated.

**Comment by yang-wang11** [comment]
@kayrus I'm not trying to rush you, but kindly look forward to your code review.

File: pkg/openstack/loadbalancer.go [hunk 4]
```diff
@@ -74,6 +74,7 @@ const (
 	ServiceAnnotationLoadBalancerFloatingSubnetTags   = "loadbalancer.openstack.org/floating-subnet-tags"
 	ServiceAnnotationLoadBalancerClass                = "loadbalancer.openstack.org/class"
 	ServiceAnnotationLoadBalancerKeepFloatingIP       = "loadbalancer.openstack.org/keep-floatingip"
+	ServiceAnnotationLoadBalancerFloatingIP           = "loadbalancer.openstack.org/floating-ip"
```
Comment (kayrus) [inline pkg/openstack/loadbalancer.go:77]: in order to comply with the previous naming, I'd suggest to use `loadbalancer.openstack.org/floatingip`
in future for dual stack we can use `loadbalancer.openstack.org/floatingip-v6`

File: pkg/openstack/loadbalancer.go [hunk 5]
```diff
@@ -431,6 +432,19 @@ func getSecurityGroupName(service *corev1.Service) string {
 	return securityGroupName
 }
 
+func getLoadBalancerIP(service *corev1.Service) string {
+	if annotations := service.Annotations; annotations != nil {
```
Comment (kayrus) [inline pkg/openstack/loadbalancer.go:436]: we can skip this check. `if val, ok := annotations[ServiceAnnotationLoadBalancerFloatingIP]; ok && val != "" {` is enough.

**Review by kayrus — COMMENTED** [review]
@yang-wang11 apologies that you had to wait. See my comments. As for the new `ensureFloatingIP` logic, give me a bit more time.

**Comment by kayrus** [comment]
@yang-wang11 Unfortunately I have doubts about the proposed `ensureFloatingIP` logic. I'm afraid the reconciliation port of the code is too fragile, and it requires a wider refactoring.

A couple of examples:

https://github.com/kubernetes/cloud-provider-openstack/blob/d6f294d0a0bdb53e2dd71023be007ef7b26d4834/pkg/openstack/loadbalancer.go#L965-L968

why would we fail in this case? What if the loadbalancer has a `service.beta.kubernetes.io/openstack-internal-load-balancer` flag?

There are also a bunch of minor concerns, which combination in result raise other questions.

Basically this is the reason why the #2377 PR is still in draft.

IMO the first step that we need to perform is to extend the mock API with Octavia support and add test cases which include a collection of different annotation combinations for services.

@yang-wang11 how critical is the issue for you? Does it occur only on svc update only?

File: pkg/openstack/loadbalancer.go (hunk 5 above)
Comment (yang-wang11) [reply pkg/openstack/loadbalancer.go:436]: Yes, it is used to avoid panic if there is no annotation. adopt your suggestion, it can cover my concern. 
https://go.dev/play/p/RDerITo3dqp?v=goprev

**Comment by yang-wang11** [comment]
> @yang-wang11 Unfortunately I have doubts about the proposed `ensureFloatingIP` logic. I'm afraid the reconciliation port of the code is too fragile, and it requires a wider refactoring.
> 
> An example:
> 
> https://github.com/kubernetes/cloud-provider-openstack/blob/d6f294d0a0bdb53e2dd71023be007ef7b26d4834/pkg/openstack/loadbalancer.go#L965-L968
> 
> why would we fail in this case? What if the loadbalancer has a `service.beta.kubernetes.io/openstack-internal-load-balancer` flag? There is no indication that this loadbalancer is going to be updated, therefore there is no reason to return an error.
> 
> There are also a bunch of minor concerns, which combination in result raises other questions.
> 
> Basically this is the reason why the #2377 PR is still in draft.
> 
> IMO the first step that we need to perform is to extend the mock API with Octavia support and add test cases which include a collection of different annotation combinations for services.
> 
> @yang-wang11 how critical is the issue for you? Does it occur only on svc update only?

@kayrus 
I'm unsure if this change belongs to a refactor or not, the thing I did is combine similar logic. It aims to eliminate the if-else logic.
As we discussed in the Slack channel, the issue that occurred from the cluster was the update case, but not sure if there are others. we hope we can fix this issue at the earliest convenience.

**Comment by kayrus** [comment]
@yang-wang11 can you resolve the conflicts?

File: docs/openstack-cloud-controller-manager/expose-applications-using-loadbalancer-type-service.md [hunk 6]
```diff
@@ -134,6 +134,12 @@ Request Body:
 
   If 'true', the floating IP will **NOT** be deleted. Default is 'false'.
 
+- `loadbalancer.openstack.org/floating-ip`
+
+  This annotation is utilized to specify the floating IP for the load balancer. 
+
+  If set, both newly created and pre-existing load balancers will utilize it. Furthermore, its precedence is higher than that of `Spec.LoadBalanccerIP`.
```
Comment (kayrus) [inline docs/openstack-cloud-controller-manager/expose-applications-using-loadbalancer-type-service.md:141]:
```suggestion
  If set, both newly created and pre-existing load balancers will utilize it. Furthermore, its precedence is higher than that of `Spec.LoadBalancerIP`.
```
